### DISCUSSION TITLE
Change the serialization lib to improve compiling performances

### DISCUSSION THREAD

**Discussion by PatchByte** [otto-link/Hesiod#104] [Q&A] [answered]
Just curious...

![grafik](https://github.com/otto-link/Hesiod/assets/69595010/07904267-aba5-4144-82d9-ba2cbc6a14a4)

**Comment by PatchByte** [comment]
I mean its compiling, but like it takes long...

**Comment by otto-link** [comment, marked as answer]
Indeed, it's probably due to the serialization library I'm using (https://github.com/USCiLab/cereal). You can comment out this line of the `CMakeLists.txt` to accelerate things:
```
set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -DUSE_CEREAL")
```
Side effect: you won't be able to load/save your node network in the GUI.

**Reply by PatchByte** [reply]
Ah thank you for your quick answer, could we implement a different serialization lib?

**Reply by otto-link** [reply]
Sure, but so far it's not my highest priority. Cereal is header-only so that could explain the resource-intensive compiling... Are you aware of a good alternative?

**Reply by PatchByte** [reply]
I personally would write a base class such as `BaseSerializable` that has 2 methods such as 
```
virtual DataType Serialize() { return DataType(); }
virtual bool Deserialize(DataType& Data) { return false; }
```

in this case we could take a library such as [nlohmann/json](https://github.com/nlohmann/json) or [DaveGamble/cJSON](https://github.com/DaveGamble/cJSON). 

So in this case this would result in:
```
virtual nlohmann::json Serialize() 
{
 nlohmann::json dataSerialized = nlohmann::json();
 dataSerialized["field_x"] = this->field_x;
 dataSerialized["field_y"] = this->field_y;
}

virtual bool Deserialize(nlohmann::json& Data)
{
 this->field_x = Data["field_x"].get<std::string>();
 this->field_y = Data["field_y"].get<int>();
}
```

**Reply by PatchByte** [reply]
I think this system is better than the current system, especially for the compile time and implementability.

**Reply by PatchByte** [reply]
If you'd like I can work on a PR

**Reply by otto-link** [reply]
We're on the same page for the serialization method, it is actually handled like this with `cereal` (in this case it's a unique method for both serialize and deserialize), for instance:
```
  template <class Archive> void serialize(Archive &ar)
  {
    ar(cereal::make_nvp("id", this->id), cereal::make_nvp("attr", this->attr));
  }
```
For the lib choice, make sure it properly handles pointers to polymorphic objects (e.g. `std::unique_ptr<Base> p( new Derived() )`) since it is widely used in the class structure, and also STL container like `std::vector` and `std::map`.

I re-created an issue #105

**Reply by PatchByte** [reply]
Alright I will get started soon

**Reply by otto-link** [reply]
Thanks a lot for contributing!

**Reply by PatchByte** [reply]
Yeah I really like this project. Hopefully we can maybe offer an open source alternative to projects such as gaea and so on!
